### PR TITLE
Add response activity item timer

### PR DESCRIPTION
**Pull request by DimitrySakalouski**
Implementation according to [https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/188547100/Items+Taking+activities+Checklist](url)
and [https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/31785052/EP-4+-+Items ](url)

### CODE REVIEW

File: src/entities/activity/model/mappers.ts [hunk 1]
```diff
@@ -42,7 +43,7 @@ function mapToDrawing(dto: DrawingItemDto): ActivityItem {
       imageUrl: dto.responseValues.drawingExample,
       backgroundImageUrl: dto.responseValues.drawingBackground,
     },
-    timer: dto.config.timer,
+    timer: dto.timer && getMsFromSeconds(dto.timer),
```
Comment (BamMironov) [inline src/entities/activity/model/mappers.ts]: Please create `mapTimer` function.

File: src/features/pass-survey/lib/context/index.ts [hunk 2]
```diff
@@ -0,0 +1 @@
+export { default as ActivityItemIdentityContext } from './ActivityItemIdentityContext';
```
Comment (BamMironov) [inline src/features/pass-survey/lib/context/index.ts:1]: Please rename the folder name: context => contexts

File: src/features/pass-survey/lib/context/ActivityItemIdentityContext.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+import { createContext } from 'react';
+
+type ActivityItemIdentifiers = {
```
Comment (BamMironov) [inline src/features/pass-survey/lib/context/ActivityItemIdentityContext.ts:3]: Please rename to ActivityIdentifiers

**Review by BamMironov — CHANGES_REQUESTED** [review]

File: src/shared/lib/timers/AppSecondTimer.ts [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  start(): void {
+    super.start();
+    this.setTimer();
+  }
+
+  setTimer() {
+    this.timerId = setInterval(() => {
+      if (this.isTimerFinished()) {
+        this.onSecondPass(this.timerId);
```
Comment (anq83) [inline src/shared/lib/timers/AppSecondTimer.ts:38]: please work on namings, not clear by name what t does

**Review by anq83 — CHANGES_REQUESTED** [review]

File: src/features/pass-survey/lib/context/index.ts (hunk 2 above)
Comment (DimitrySakalouski) [reply src/features/pass-survey/lib/context/index.ts:1]: done

File: src/features/pass-survey/lib/context/ActivityItemIdentityContext.ts (hunk 3 above)
Comment (DimitrySakalouski) [reply src/features/pass-survey/lib/context/ActivityItemIdentityContext.ts:3]: done

File: src/features/pass-survey/lib/hooks/useActivityStorageRecord.ts [hunk 5]
```diff
@@ -16,11 +16,17 @@ export type Answer = PipelineItemAnswer['value'];
 
 export type Answers = Record<string, Answer>;
 
+type Timer = {
+  progress: number;
+  timerId: number;
```
[the remark this answers is not in the document]
Comment (DimitrySakalouski) [reply src/features/pass-survey/lib/hooks/useActivityStorageRecord.ts:21]: done

File: src/features/pass-survey/model/hooks/useActivityState.ts [hunk 6]
```diff
@@ -83,13 +83,46 @@ function useActivityState({
...
+    });
+  }
+
+  function removeTimer() {
+    if (!activityStorageRecord) {
+      return;
+    }
+
+    if (activityStorageRecord.timer?.timerId) {
+      clearInterval(activityStorageRecord.timer?.timerId);
```
[the remark this answers is not in the document]
Comment (DimitrySakalouski) [reply src/features/pass-survey/model/hooks/useActivityState.ts:107]: done

File: src/shared/lib/timers/AppSecondTimer.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import TimerBase from './TimerBase';
+
+class AppSecondTimer extends TimerBase {
```
[the remark this answers is not in the document]
Comment (DimitrySakalouski) [reply src/shared/lib/timers/AppSecondTimer.ts:3]: done

File: src/entities/activity/model/timerMapper.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { getMsFromSeconds } from '@app/shared/lib';
+
+const mapTimerValue = (dtoTimer: number | null) => {
+  if (dtoTimer) {
+    return getMsFromSeconds(dtoTimer);
+  }
+
+  return dtoTimer;
```
[the remark this answers is not in the document]
Comment (DimitrySakalouski) [reply src/entities/activity/model/timerMapper.ts:8]: done

File: src/features/pass-survey/ui/ActivityStepper.tsx [hunk 9]
```diff
@@ -131,98 +135,102 @@ function ActivityStepper({
...
-
-            return (
-              <XStack flex={1} key={index} alignItems="center">
-                <>
-                  {pipelineItem.type === 'Tutorial' && (
-                    <TutorialViewerItem
-                      {...pipelineItem.payload}
-                      ref={tutorialViewerRef}
-                    />
+    <ActivityItemIdentityContext.Provider
```
[the remark this answers is not in the document]
Comment (DimitrySakalouski) [reply src/features/pass-survey/ui/ActivityStepper.tsx:138]: done

File: src/shared/lib/timers/AppSecondTimer.ts (hunk 4 above)
Comment (DimitrySakalouski) [reply src/shared/lib/timers/AppSecondTimer.ts:38]: done

File: src/features/pass-survey/model/hooks/useActivityState.ts [hunk 10]
```diff
@@ -83,13 +83,41 @@ function useActivityState({
...
 
+  function setTimer(step: number, progress: number) {
+    if (!activityStorageRecord) {
+      return;
+    }
+
+    upsertActivityStorageRecord({
+      ...activityStorageRecord,
+
+      timers: {
```
Comment (BamMironov) [inline src/features/pass-survey/model/hooks/useActivityState.ts:94]: Although it is not important at the moment, I would recommend changing it right now this way:
...activityStorageRecord.timers,
[step]: progress,

File: src/features/pass-survey/ui/Timer.tsx [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    activityId,
+    eventId,
+  });
+
+  const progressDone = useMemo(() => {
+    if (
+      activityStorageRecord?.timers &&
+      Object.keys(activityStorageRecord?.timers).length
+    ) {
+      return activityStorageRecord?.timers[activityStorageRecord.step];
```
Comment (BamMironov) [inline src/features/pass-survey/ui/Timer.tsx:74]: return activityStorageRecord.timers[activityStorageRecord.step] ?? 0;

File: src/features/pass-survey/ui/Timer.tsx [hunk 12]
```diff
@@ -0,0 +1,135 @@
+/* eslint-disable react-hooks/exhaustive-deps */
```
Comment (BamMironov) [inline src/features/pass-survey/ui/Timer.tsx:1]: Please delete this

**Review by BamMironov — CHANGES_REQUESTED** [review]

**Review by BamMironov — APPROVED** [review]

File: src/widgets/survey/ui/FlowElementSwitch.tsx [hunk 13]
```diff
@@ -29,20 +32,22 @@ function FlowElementSwitch({
...
-            {...payload}
-            idleTimer={event.timers.idleTimer}
-            timer={event.timers.timer}
-            entityStartedAt={entityStartedAt}
-            onClose={onClose}
-            onFinish={onComplete}
-          />
-        </Box>
+            <ActivityStepper
+              {...payload}
```
Comment (anq83) [inline src/widgets/survey/ui/FlowElementSwitch.tsx:42]: No need in payload anymore as you use context now

File: src/features/pass-survey/ui/ActivityStepper.tsx [hunk 14]
```diff
@@ -107,11 +108,13 @@ function ActivityStepper({
...
   const onNext = (nextStep: number) => {
+    removeTimer(currentStep);
     setCurrentStep(nextStep);
     restartIdleTimer();
   };
 
   const onBack = (nextStep: number) => {
+    removeTimer(currentStep);
     setCurrentStep(nextStep);
     restartIdleTimer();
```
Comment (anq83) [inline src/features/pass-survey/ui/ActivityStepper.tsx:119]: the same

**Review by anq83 — CHANGES_REQUESTED** [review]

**Review by anq83 — APPROVED** [review]

**Pull request merged**